python/plugins/processing/algs/qgis/CreateAtlasGrid.py

```diff
@@ -0,0 +1,197 @@
+
+    def group(self):
+        return self.tr('Cartography')
+
+    def groupId(self):
+        return 'cartography'
+
+    def __init__(self):
+        super().__init__()
+
+    def initAlgorithm(self, config=None):
+
+        param = QgsProcessingParameterLayout(self.LAYOUT, 'Print layout')
+        self.addParameter(param)
+        map_param = QgsProcessingParameterLayoutItem(self.MAP, 'Map item', parentLayoutParameterName=self.LAYOUT,
+                                                     itemType=QgsLayoutItemRegistry.LayoutMap)
+        self.addParameter(map_param)
+        self.addParameter(QgsProcessingParameterFeatureSource(self.COVERAGE_LAYER,
+                                                              self.tr('Coverage layer'),
+                                                              [QgsProcessing.TypeMapLayer]))
+        self.addParameter(QgsProcessingParameterDistance(self.COVERAGE_BUFFER,
+                                                         self.tr('Coverage buffer'), parentParameterName='COVERAGE_LAYER',
+                                                         defaultValue=0.0))
+        self.addParameter(QgsProcessingParameterBoolean(self.ONLY_ON_FEATURES,
+                                                        self.tr('Only on intersected features'),
+                                                        defaultValue=True))
+        self.addParameter(QgsProcessingParameterFeatureSink(self.OUTPUT,
+                                                            self.tr('Grid'),
+                                                            QgsProcessing.TypeVectorPolygon))
+
+    def prepareAlgorithm(self, parameters, context, feedback):
+
+        layout = self.parameterAsLayout(parameters, self.LAYOUT, context)
+        if layout is None:
+            raise QgsProcessingException('Cannot find layout with name "{}"'.format(parameters[self.LAYOUT]))
+
+        map = self.parameterAsLayoutItem(parameters, self.MAP, context, layout)
+        if map is None:
+            raise QgsProcessingException('Cannot find matching map item with ID {}'.format(parameters[self.MAP]))
+
+        self.mapheight = map.extent().height()
+        self.mapwidth = map.extent().width()
+
+        self.coverage_layer = self.parameterAsVectorLayer(
+            parameters,
+            self.COVERAGE_LAYER,
+            context
+        )
+        coverage_distance = self.parameterAsDouble(
+            parameters,
+            self.COVERAGE_BUFFER,
+            context
+        )
+        self.only_on_features = self.parameterAsBool(
+            parameters,
+            self.ONLY_ON_FEATURES,
+            context
+        )
+        self.coverage = self.coverage_layer.sourceExtent().buffered(coverage_distance)
+
+        self.total = (self.coverage.height() / self.mapheight) * (self.coverage.width() / self.mapwidth)
+
+        mapcrs = map.crs()
+        laycrs = self.coverage_layer.crs()
+
+        if mapcrs != laycrs:
```

Comment:
better to just handle this for users. I would always use the map crs, as this simplifies things.

To do this:
1. use .setDestinationCrs(self.mapcrs) on your requests to the layer to automatically reproject them to the map crs 
2. reproject the layer extent when calculating self.coverage to the map crs
3. create your sink using the map crs

(and better add a unit test covering this situation)